Add PCL port

UnaNancyOwen opened the pull request:
Add Point Cloud Library (PCL)

UnaNancyOwen committed:
Add PCL port

Add Point Cloud Library (PCL)

msftclas commented:
@__UnaNancyOwen__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla.microsoft.com.

It will cover your contributions to all Microsoft-managed open source projects.
_Thanks,
Microsoft Pull Request Bot_

msftclas commented:
__@UnaNancyOwen__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, Microsoft Pull Request Bot_

seanlis commented:
FIND_PACKAGE(PCL) failed.
It seems that [PCLConfig.cmake.in#L708-L715](https://github.com/PointCloudLibrary/pcl/blob/pcl-1.8.0/PCLConfig.cmake.in#L708-L715) shoud have been changed, since the configuration files are installed in share/pcl.

UnaNancyOwen commented:
@seanlis Thanks, I will fix it.

jasjuang commented:
@UnaNancyOwen Can you let me know what I am doing wrong.

``````
git clone https://github.com/UnaNancyOwen/vcpkg
cd vcpkg 
git checkout add_pcl
bootstrap-vcpkg.bat
vcpkg install pcl
Error: while loading pcl: no such file or directory
``````

seanlis commented on ports/pcl/portfile.cmake:
```diff
@@ -0,0 +1,71 @@
+    #PREFER_NINJA
+    OPTIONS
+        # BUILD
+        -DBUILD_2d=OFF # The pcl-2d import library is not generated.
```
Is it possible to enable 2d module? Several other modules depend on it.
Because it's a header only library, I think maybe the dll could be deleted.
I did the following change on my machine, got no error so far.
```diff
diff --git a/ports/pcl/portfile.cmake b/ports/pcl/portfile.cmake
index 6285d88c..e885eaa4 100644
--- a/ports/pcl/portfile.cmake
+++ b/ports/pcl/portfile.cmake
@@ -45,7 +45,6 @@ vcpkg_configure_cmake(
     #PREFER_NINJA
     OPTIONS
         # BUILD
-        -DBUILD_2d=OFF # The pcl-2d import library is not generated.
         -DBUILD_surface_on_nurbs=ON
         -DBUILD_tools=OFF
         # PCL
@@ -64,6 +63,7 @@ vcpkg_configure_cmake(

 vcpkg_install_cmake()

+file(REMOVE ${CURRENT_PACKAGES_DIR}/bin/pcl_2d_release.dll ${CURRENT_PACKAGES_DIR}/debug/bin/pcl_2d_debug.dll)
 file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
 file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/share)
 file(COPY ${SOURCE_PATH}/LICENSE.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/pcl)
```

UnaNancyOwen replied:
Thank you for your feedback.
You're right. I will fix it soon.

UnaNancyOwen committed:
Fix to enable pcl_2d module

Fix to enable pcl_2d module.
It's a header only library.

UnaNancyOwen commented:
@jasjuang Sorry, I'm not familiar about Git operation.
Please clone latest version (master/HEAD) of Microsoft/vcpkg, and copy this port to latest vcpkg.

UnaNancyOwen commented:
Please merge #1329 to fix QHull port.

UnaNancyOwen committed:
Update PCL 1.8.1 RC1

Update PCL 1.8.1 RC1

KindDragon commented:
PCLConfig.cmake contains link to release libraries, cannot build debug configuration with PCL

jasjuang commented:
@UnaNancyOwen I just tried it and I wasn't able to compile it. I had to add the below lines right under `vcpkg_install_cmake()` in portfile.cmake

``````
vcpkg_fixup_cmake_targets(CONFIG_PATH "cmake")

vcpkg_copy_pdbs()
``````

in order for it to compile. But the problem is when I attempt to use PCL in my code with the standard

``````
find_package( PCL 1.8 REQUIRED )

target_include_directories(${PROJECT_NAME} SYSTEM PUBLIC
  ${PCL_INCLUDE_DIRS}
)

target_link_libraries(${PROJECT_NAME} ${PCL_LIBRARIES})
``````

I get the below error when running cmake 

``````
CMake Error at C:/dev/vcpkg/installed/x64-windows/share/pcl/PCLConfig.cmake:44 (message):
  PCL can not be found on this machine
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows/share/pcl/PCLConfig.cmake:768 (pcl_report_not_found)
  apps/SampleExample/CMakeLists.txt:10 (find_package)


-- Configuring incomplete, errors occurred!
``````

It's not obvious to me on how to fix it. Could you look into it?

Voskrese commented:
build ok
PCL 1.8.1 RC1 - x86 Dynamic

UnaNancyOwen commented:
find_library() always detect debug libraries of FLANN.
(It is used in share/pcl/PCLConfig.cmake.)
The cause of this problem is the suffix doesn't exist in the debug library of FLANN.
I propose to add the CMAKE_DEBUG_POSTFIX to FLANN port.

UnaNancyOwen committed:
Enable Qt option

Enable Qt option

UnaNancyOwen commented:
The own project generated using CMake is always to link following libraries of release version.
(freetype, zlib, expatm, libpng16, tiff, hdf5, msmpi, hdf5_hl, libxml2, jsoncpp)
These should link the debug library in debug configuration.
Do you have any ideas to solve this problem?
![debug](https://user-images.githubusercontent.com/816705/28227655-ed684250-6915-11e7-9a20-2d6b7a799a83.png)

UnaNancyOwen committed:
Fix find_flann.patch

Fix find_flann.patch according to this change (PR#1452).

ras0219-msft committed:
[pcl] Add qt5 dependency.

UnaNancyOwen commented on ports/pcl/CONTROL:
```diff
@@ -0,0 +1,4 @@
+Source: pcl
+Version: 1.8.1rc1-1
```
@ras0219-msft Why did you increment version?
I think that version won't be necessary to increment for now, because this port has not been merged yet.

UnaNancyOwen commented on ports/pcl/CONTROL:
```diff
@@ -1,4 +1,4 @@
-Version: 1.8.1rc1
-Build-Depends: boost, eigen3, flann, qhull, vtk, openni2
+Version: 1.8.1rc1-1
+Build-Depends: boost, eigen3, flann, qhull, vtk, openni2, qt5
```
Thanks, it's necessary for WITH_QT.

UnaNancyOwen commented on ports/pcl/portfile.cmake:
```diff
@@ -43,8 +43,8 @@ vcpkg_configure_cmake(
-        -DPCL_BUILD_WITH_BOOST_DYNAMIC_LINKING_WIN32=${CRT_LINKAGE}
-        -DPCL_BUILD_WITH_FLANN_DYNAMIC_LINKING_WIN32=${CRT_LINKAGE}
+        -DPCL_BUILD_WITH_BOOST_DYNAMIC_LINKING_WIN32=ON
+        -DPCL_BUILD_WITH_FLANN_DYNAMIC_LINKING_WIN32=ON
```
@ras0219-msft Why did you always turn on these options?

ras0219-msft replied:
We prefer that the dependencies always look like they are dynamic linked so the consumer does not need to be aware of the downstream linkage.

UnaNancyOwen replied:
These options mean to use dynamic link libraries of Boost and FLANN.
I think these options should be set to ON when triplet is dynamic (<code>x86-windows</code> or <code>x64-windows</code>), set to OFF when triplet is static (<code>x86-windows-static</code> or <code>x64-windows-static</code>).
What do you think about it?
(Should It switch these options with <code>VCPKG_LIBRARY_LINKAGE</code> instead of <code>VCPKG_CRT_LINKAGE</code>?)

ras0219-msft replied:
The trouble is that it is conceivable that users create their own triplets (`x86-windows-custom`) which specify dynamic linkage for those libraries but static linkage for PCL.

Therefore, we solve this by making all libraries _appear_ to be dynamically linked. Libraries like PCL then always request dynamic linkage and they will receive static or dynamic libs based on how the downstream library was configured.

UnaNancyOwen replied:
Sorry, I still don't understand that.

For example, If users use <code>vcpkg install pcl:x86-windows-static</code>, the dependent libraries are also built with same configuration (<code>boost:x86-windows-static</code>, <code>flann:x86-windows-static</code>).
In this case, Vcpkg to be generated and installed static link libraries of Boost and FLANN, not dynamic link libraries. And, PCL links these static libraries.
At that time, I think that these options should be turned off, because Boost and FLANN dyamic link libraries are not installed.

Is that my misunderstanding?

ras0219-msft replied:
Users can create their own triplets: http://vcpkg.readthedocs.io/en/latest/users/triplets/.

Therefore, we try to rename the produced binaries under _all_ triplets to be consistent [1]. This way, consumers can be ignorant of how their downstream libraries are produced. This works because Vcpkg guarantees only one build of the library will be available -- there are never two to choose from.

[1] https://github.com/Microsoft/vcpkg/blob/c8e26828e41ca1032dc4d6cb6660d4a5ca588d4c/ports/boost/portfile.cmake#L243-L246

UnaNancyOwen replied:
I understood that the libraries has same name in all triplets. It's no problem.
But, PCL defines several different compiler flags and preprocessor definitions depending on the targets (dynamic or static). These options should be switching for control it.

UnaNancyOwen committed:
Update PCL 1.8.1 RC2

Update PCL 1.8.1 RC2

UnaNancyOwen commented:
This problem https://github.com/Microsoft/vcpkg/pull/1283#issuecomment-315448512 is caused by the VTK port.
<code>VTK_USE_SYSTEM_*</code> options means uses dependent libraries that installed in system (installed by vcpkg), instead of dependent libraries that included in VTK.
In this case, VTK seems to link the dependent libraries of release version when debug configuration.
This is part of <code>vcpkg\buildtrees\vtk\x86-windows-dbg\CMakeCache.txt</code>.
For example, This means that vtkIOImage will link libpng.
```
//Dependencies for the target
vtkIOImage_LIB_DEPENDS:STATIC= ... ;general;C:/vcpkg/installed/x86-windows/lib/libpng16.lib;general;C:/vcpkg/installed/x86-windows/debug/lib/libpng16d.lib; ...;
```
But, It should be output as follows.
```
//Dependencies for the target
vtkIOImage_LIB_DEPENDS:STATIC= ... ;optimized;C:/vcpkg/installed/x86-windows/lib/libpng16.lib;debug;C:/vcpkg/installed/x86-windows/debug/lib/libpng16d.lib; ...;
```
This problem can be easily resolved by using dependency libraries that included in VTK.
It's easy. It only changes to off these options.
@ras0219-msft Do you have any special reasons that use <code>VTK_USE_SYSTEM_*</code> option?

ras0219-msft commented:
We need to use the vcpkg versions of libraries in order to enable:
1. Static linking. If multiple ports provide conflicting symbols (even internally), you'll get random behavior at link time and (probably) runtime crashes.
2. Security. If libpng's parser has a CVE, we need to be able to update libpng to fix it. This only works when other libraries are linking against the system copy of libpng.
3. Correctness. As a user, if I link against libpng and manipulate some global settings (let's say there was one to change default search paths), it should be reflected in any library that uses png. Additionally, if VTK has any APIs that use structures defined by libpng, I _must_ use the same exact libpng that VTK was built against in order to have ABI compatibility.

UnaNancyOwen commented:
@ras0219-msft I see. Do you have any good ideas to solve this problem https://github.com/Microsoft/vcpkg/pull/1283#issuecomment-315448512 ?

UnaNancyOwen commented:
@ras0219-msft I sent a new pull request #1568 that to solve this problem. Please review it. Thanks!

UnaNancyOwen commented:
I think all problems has been resolved.
This port is able to merge when PCL 1.8.1 is officially released.
If you find more problems, please tell me.
Thank you everyone!

jasjuang commented:
@UnaNancyOwen I am trying to test but flann would not build for me. I am on cmake 3.9 and vs2015 and trying to build x64. I will file an separate issue. After that is resolved I will come back and test this.

UnaNancyOwen committed:
Update PCL 1.8.1

Update PCL 1.8.1

UnaNancyOwen commented:
PCL 1.8.1 has been officially released today. I updated this port to PCL 1.8.1.
Please review it. And, If it have no problems, this port able to merge!
Thanks,

UnaNancyOwen commented:
I found one problem when using <code>vcpkg integrate install</code>.

Usually, the header files of PCL are installed in the following directory structures.
```cmake
vcpkg\installed\x86-windows\include\pcl-1.8\pcl\<header.h>
```
And, the users include header files in source code as follows.
```cpp
#include <pcl/header.h>
```

The <code>vcpkg integrate install</code> command makes it can use libraries that installed by vcpkg in all Visual C++ projects. It sets the include directories setting to <code>vcpkg\installed\x86-windows\include</code>.
But, This directory doesn't have PCL header files.

I think that the PCL header files should be installed in the following directory structure.
```cmake
#vcpkg\installed\x86-windows\include\pcl-1.8\pcl\<header.h>
vcpkg\installed\x86-windows\include\pcl\<header.h>
```

I will fix this.

BTW:
VTK port have also same problem. Can you solve this problem in VTK port?

UnaNancyOwen committed:
Change directory structure to install the header files

Fix bug when using "vcpkg integrate install" command by change directory
structure to install the header files.

UnaNancyOwen commented:
This problem https://github.com/Microsoft/vcpkg/pull/1283#issuecomment-321610342 was fixed.
Please review it. Thanks,
(But, the VTK port hasn't been fixed yet. Please manually add the VTK headers path to "additional include directories" when using "vcpkg integrate install" command.)

alexkaratarakis commented:
Thanks for the solid work! #1601 has also been merged, so the `vcpkg integrate install` workaround should not be needed.
